node/pkg/p2p/ccq_p2p.go

```diff
@@ -0,0 +1,296 @@
+	"github.com/certusone/wormhole/node/pkg/query"
+	ethcrypto "github.com/ethereum/go-ethereum/crypto"
+	"github.com/prometheus/client_golang/prometheus"
+	"github.com/prometheus/client_golang/prometheus/promauto"
+	"google.golang.org/protobuf/proto"
+
+	gossipv1 "github.com/certusone/wormhole/node/pkg/proto/gossip/v1"
+
+	"github.com/libp2p/go-libp2p"
+	dht "github.com/libp2p/go-libp2p-kad-dht"
+	pubsub "github.com/libp2p/go-libp2p-pubsub"
+	"github.com/libp2p/go-libp2p/core/crypto"
+	"github.com/libp2p/go-libp2p/core/host"
+	"github.com/libp2p/go-libp2p/core/peer"
+	"github.com/libp2p/go-libp2p/core/protocol"
+	"github.com/libp2p/go-libp2p/core/routing"
+	libp2ptls "github.com/libp2p/go-libp2p/p2p/security/tls"
+	libp2pquic "github.com/libp2p/go-libp2p/p2p/transport/quic"
+	"go.uber.org/zap"
+)
+
+var (
+	ccqP2pMessagesSent = promauto.NewCounter(
+		prometheus.CounterOpts{
+			Name: "wormhole_ccqp2p_broadcast_messages_sent_total",
+			Help: "Total number of ccq p2p pubsub broadcast messages sent",
+		})
+	ccqP2pMessagesReceived = promauto.NewCounterVec(
+		prometheus.CounterOpts{
+			Name: "wormhole_ccqp2p_broadcast_messages_received_total",
+			Help: "Total number of ccq p2p pubsub broadcast messages received",
+		}, []string{"type"})
+)
+
+type ccqP2p struct {
+	logger *zap.Logger
+
+	h            host.Host
+	th_req       *pubsub.Topic
+	th_resp      *pubsub.Topic
+	sub          *pubsub.Subscription
+	allowedPeers map[string]struct{}
+}
+
+func newCcqRunP2p(
+	logger *zap.Logger,
+	allowedPeersStr string,
+) *ccqP2p {
+	l := logger.With(zap.String("component", "ccqp2p"))
+	allowedPeers := make(map[string]struct{})
+	for _, peerID := range strings.Split(allowedPeersStr, ",") {
+		if peerID != "" {
+			l.Info("will allow requests from peer", zap.String("peerID", peerID))
+			allowedPeers[peerID] = struct{}{}
+		}
+	}
+
+	return &ccqP2p{
+		logger:       l,
+		allowedPeers: allowedPeers,
+	}
+}
+
+func (ccq *ccqP2p) run(
+	ctx context.Context,
+	priv crypto.PrivKey,
+	gk *ecdsa.PrivateKey,
+	networkID string,
+	bootstrapPeers string,
+	port uint,
+	signedQueryReqC chan<- *gossipv1.SignedQueryRequest,
+	queryResponseReadC <-chan *query.QueryResponsePublication,
+	errC chan error,
+) error {
+	var err error
+
+	components := DefaultComponents()
+	if components == nil {
+		return fmt.Errorf("components is not initialized")
+	}
+	components.Port = port
+
+	ccq.h, err = libp2p.New(
+		// Use the keypair we generated
+		libp2p.Identity(priv),
+
+		// Multiple listen addresses
+		libp2p.ListenAddrStrings(
+			components.ListeningAddresses()...,
+		),
+
+		// Enable TLS security as the only security protocol.
+		libp2p.Security(libp2ptls.ID, libp2ptls.New),
+
+		// Enable QUIC transport as the only transport.
+		libp2p.Transport(libp2pquic.NewTransport),
+
+		// Let's prevent our peer from having too many
+		// connections by attaching a connection manager.
+		libp2p.ConnectionManager(components.ConnMgr), // TODO: Can we use the same connection manager?
```

Comment:
I think that each p2p instance should be using its own connection manager otherwise it could prune each other's connections

Reply:
The code is actually creating a separate `components` object at line 87, which means the connection manager is not shared. I'll remove that todo with my next PR.